clarity/contracts/arkadiko-auction-engine-v4-1.clar

```diff
@@ -0,0 +1,396 @@
+
+        ;; Burn USDA, get collateral token
+        (try! (as-contract (contract-call? .arkadiko-dao burn-token .usda-token usda-to-use (as-contract tx-sender))))
+        (try! (as-contract (contract-call? vault-manager redeem-auction-collateral ft token-string reserve collateral-sold (as-contract tx-sender)))) 
+
+        ;; Deposit collateral token
+        (try! (as-contract (contract-call? liquidation-rewards add-reward block-height token-is-stx ft collateral-sold)))
+
+        ;; Update auction
+        (map-set auctions { id: auction-id } (merge auction {
+          total-collateral-sold: (+ (get total-collateral-sold auction) collateral-sold),
+          total-debt-burned: (+ (get total-debt-burned auction) usda-to-use)
+        }))
+
+        true
+      )
+    )
+
+    ;; Try to finalize
+    (try! (finalize-liquidation auction-id vault-manager coll-type))
+
+    (print { type: "auction", action: "burn-usda", data: auction })
+    (ok usda-to-use)
+  )
+)
+
+;; @desc finalize liquidation so rest of collateral becomes available to vault owner
+;; @param auction-id; the auction
+;; @param vault-manager; vault manager
+;; @param coll-type; collateral types
+(define-public (finalize-liquidation
+  (auction-id uint)
+  (vault-manager <vault-manager-trait>)
+  (coll-type <collateral-types-trait>)
+)
+  (let (
+    (auction (get-auction-by-id auction-id))
+  )
+    (asserts! (not (shutdown-activated)) (err ERR-EMERGENCY-SHUTDOWN-ACTIVATED))
+    (asserts! (is-eq (contract-of vault-manager) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "freddie"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of coll-type) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "collateral-types"))) (err ERR-NOT-AUTHORIZED))
+
+    ;; Sell off DIKO if needed
+    (try! (sell-diko auction-id))
+
+    ;; Do not continue if auction still open
+    (asserts! (not (get-auction-open auction-id)) (ok false))
+
+    ;; Finalize liquidation
+    (try! (as-contract (contract-call?
+      vault-manager
+      finalize-liquidation
+      (get vault-id auction)
+      (- (get collateral-amount auction) (get total-collateral-sold auction))
+      coll-type
+    )))
+    (map-set auctions { id: auction-id } (merge (get-auction-by-id auction-id) { ended-at: block-height }))
+  
+    (print { type: "auction", action: "finalize-liquidation", data: auction })
+    (ok true)
+  )
+)
+
+;; ---------------------------------------------------------
+;; Helpers
+;; ---------------------------------------------------------
+
+;; If all collateral sold but still debt to raise,
+;; DIKO is minted and swapped to USDA to pay off debt
+(define-private (sell-diko (auction-id uint))
+  (let (
+    (auction (get-auction-by-id auction-id))
+    (all-collateral-sold (is-eq (get collateral-amount auction) (get total-collateral-sold auction)))
+    (debt-left (- (get debt-to-raise auction) (get total-debt-burned auction)))
```

Comment:
Q/REQ: maybe we can change this for `(>= (get total-debt-burned auction) (get debt-to-raise auction))`. I don't know if it can ever happen, but if this would be a negative number, this throws an error and as a result, an auction can never be finalised